src/python/grpcio/grpc/experimental/aio/_call.py

```diff
@@ -56,207 +81,331 @@ def __init__(self,
-          response_deserializer: Deserializer used for parsing the reponse.
-          call_cancel_status: A cygrpc.AioCancelStatus used for giving a
-            specific error when the RPC is canceled.
-        """
 
-        self._state = _RpcState.ONGOING
-        self._exception = None
-        self._response = None
-        self._code = grpc.StatusCode.UNKNOWN
-        self._details = None
-        self._initial_metadata = None
-        self._trailing_metadata = None
-        self._call = call
-        self._call_cancel_status = call_cancel_status
-        self._response_deserializer = response_deserializer
+class Call(_base_call.Call):
+    _loop: asyncio.AbstractEventLoop
+    _code: grpc.StatusCode
+    _status: Awaitable[cygrpc.AioRpcStatus]
+    _initial_metadata: Awaitable[MetadataType]
+    _cancellation_future: asyncio.Future
 
-    def __del__(self):
-        self.cancel()
+    def __init__(self) -> None:
+        self._loop = asyncio.get_event_loop()
+        self._code = None
+        self._status = self._loop.create_future()
+        self._initial_metadata = self._loop.create_future()
+        self._cancellation_future = self._loop.create_future()
 
     def cancel(self) -> bool:
-        """Cancels the ongoing RPC request.
+        """Virtual cancellation method.
 
-        Returns:
-          True if the RPC can be canceled, False if was already cancelled or terminated.
+        The implementation of this method needs to pass the cancellation reason
+        into self._cancellation_future, using `set_result` instead of
+        `set_exception`.
         """
-        if self.cancelled() or self.done():
-            return False
-
-        code = grpc.StatusCode.CANCELLED
-        self._call_cancel_status.cancel(
-            _common.STATUS_CODE_TO_CYGRPC_STATUS_CODE[code],
-            details=Call._cancellation_details)
-        self._call.cancel()
-        self._details = Call._cancellation_details
-        self._code = code
-        self._state = _RpcState.CANCELLED
-        return True
+        raise NotImplementedError()
 
     def cancelled(self) -> bool:
-        """Returns if the RPC was cancelled.
-
-        Returns:
-          True if the requests was cancelled, False if not.
-        """
-        return self._state is _RpcState.CANCELLED
-
-    def running(self) -> bool:
-        """Returns if the RPC is running.
-
-        Returns:
-          True if the requests is running, False if it already terminated.
-        """
-        return not self.done()
+        return self._cancellation_future.done(
+        ) or self._code == grpc.StatusCode.CANCELLED
 
     def done(self) -> bool:
-        """Returns if the RPC has finished.
+        return self._status.done()
 
-        Returns:
-          True if the requests has finished, False is if still ongoing.
-        """
-        return self._state is not _RpcState.ONGOING
+    def add_callback(self, unused_callback) -> None:
```

Comment:
I'm torn about this, this callback design is not ideal. I have seen many users asking how to use it, because it only accepts functions like `Callable[[None], None]`. If users are not familiar with factory pattern, then they will have a hard time using it.

On the contrary, the `Future.add_done_callback` supplies the future itself to callbacks, which means the callback have directly access to the `grpc.aio.Call` object. It would be much easier to use.

However, this method is supported in our existing API, I'm not sure if we can simply get rid of it.

Reply:
Absolutely we can get rid of it. This is a new API after all. Better to have one consistent way to add a done callback than two potentially incompatible ways.

Reply:
Method removed in https://github.com/grpc/grpc/pull/21232/commits/fa4eb94ea22aa6370a6f572cd3e8a01a75032b27